### PR TITLE
send websocket messages without flushing, with periodic flush

### PR DESCRIPTION
**Pull request by ornicar**
not used yet, just implementing the behaviour

### CODE REVIEW

File: src/main/scala/util/RunPeriodically.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    synchronized:
+      counter += 1
+      if counter >= counterMax then unsafeFlush()
+
+  private def flush(): Unit = synchronized { unsafeFlush() }
+
+  private def unsafeFlush(): Unit =
+    if counter > 0 then
+      run()
+      counter == 0
```
Comment (lenguyenthanh) [inline src/main/scala/util/RunPeriodically.scala:32]: should it be `counter = 0` 🤔

Comment (ornicar) [reply src/main/scala/util/RunPeriodically.scala:32]: Dammit discarded values!!

Comment (schlawg) [reply src/main/scala/util/RunPeriodically.scala:32]: i don't really understand what this does. i was thinking something like:

```scala
package lila.ws
package util

import io.netty.channel.*
import org.apache.pekko.actor.Cancellable
import org.apache.pekko.actor.typed.Scheduler

final class RunPeriodically()(using scheduler: Scheduler, ec: Executor):
  @volatile var flushQ = scala.collection.mutable.LinkedHashSet.empty[Channel]

  scheduler.scheduleAtFixedRate(0 seconds, 1 millisecond)(() => flushEm())

  def apply(channel: Channel) =
    synchronized:
      flushQ += channel

  private def flushEm(): Unit =
    val toFlush = synchronized:
      val f = flushQ.take(100)
      flushQ --= f
      f
    // if flushQ.size > 500 then we're drowning, panic, log something, etc.
    toFlush.foreach(_.flush())
```

just an example

Comment (schlawg) [reply src/main/scala/util/RunPeriodically.scala:32]: and we play with the numbers. could be scheduleAtFixedRate(0, 500 milliseconds) and flushQ.take(5000), etc.

**Comment by schlawg** [comment]
i don't really understand what this does. i was thinking something like:

```scala
final class FlushPeriodically()(using scheduler: Scheduler, ec: Executor):
  val flushQ = scala.collection.mutable.LinkedHashSet.empty[Channel]

  scheduler.scheduleAtFixedRate(0 seconds, 1 millisecond)(() => flushEm())

  def apply(channel: Channel) =
    synchronized: // maybe ConcurrentHashMap could avoid synchronized
      flushQ += channel

  private def flushEm(): Unit =
    val toFlush = synchronized:
      val f = flushQ.take(100)
      flushQ --= f
      f
    // if flushQ.size > 5000 then we're drowning, panic, log something, etc.
    toFlush.foreach(_.flush())
```
and then we play with the numbers.

the above doesn't do any batching, but as a framework it's easier for me to understand where to add stuff to get what we want (i.e. making sure that a channel stays in the set/map for at least so many intervals before flushing to give batching a chance, stuff like that.)

File: src/main/scala/util/RunPeriodically.scala (hunk 1 above)
Comment (schlawg) [reply src/main/scala/util/RunPeriodically.scala:32]: i don't think it really matters, the clientEmit closure seems to recreate itself with some regularity.

Comment (schlawg) [reply src/main/scala/util/RunPeriodically.scala:32]: probably because i'm reconnecting..

**Comment by schlawg** [comment]
i cannot test the code in this PR because clients just keep reconnecting with the emitToChannelWithPeriodicFlush lambda path. could scala hate me more?

```scala
  private def emitToChannel(channel: Channel): ClientEmit =
    // partial function with direct pattern matching
  private def emitToChannelWithPeriodicFlush(channel: Channel): ClientEmit =
    // closure & lambda wrapper
    msg =>
      msg match
        // pattern matching
```

**Comment by ornicar** [comment]
the client reconnects because lila-ws doesn't respond to its `null` ping with the expected `0` pong.

**Comment by schlawg** [comment]
except for adding:
```
        case ipc.ClientIn.Pong =>
          channel.writeAndFlush("0")
```
doesn't help.

> the client reconnects because lila-ws doesn't respond to its `null` ping with the expected `0` pong.

yeah, i had commented out all of the periodically stuff, changed to use writeAndFlush in both, and was left with the only difference between the two methods being one was wrapped in another layer of lambda. and swapping between them the wrapped one was still  reconnecting every few seconds. but this morning things are behaving, so maybe i only thought i was rebuilding? at least it's me that's crazy and not scalac or netty.

**Pull request closed without merging**